File: app/rcc/scripts/mavlink.js
/**
 * MAVLink Messages Parser
 *
 * Parses MAVLink protocol messages (commonly used in drones and robotics).
 *
 * MAVLink is a lightweight messaging protocol for communicating with drones,
 * ground control stations, and onboard drone components.
 *
 * INPUT FORMAT: Binary MAVLink v1 or v2 frame
 * OUTPUT ARRAY: [value1, value2, value3, ...]  (extracted from message payloads)
 *
 * Note: This parser requires Binary (Direct) decoder mode in the Project Editor.
 *       Frame delimiters are automatically removed by Serial Studio.
 *       Values are extracted based on message ID and mapped to array indices.
 */

//------------------------------------------------------------------------------
// Configuration
//------------------------------------------------------------------------------

/**
 * Total number of values in the output array.
 * This should match the number of datasets you've configured in Serial Studio.
 */
const numItems = 16;

/**
 * MAVLink protocol version to parse.
 * Set to 1 for MAVLink v1 (0xFE start marker)
 * Set to 2 for MAVLink v2 (0xFD start marker)
 */
const mavlinkVersion = 1;

/**
 * Maps MAVLink message IDs to parsing configurations.
 *
 * Common MAVLink message IDs:
 *   0:   HEARTBEAT
 *   24:  GPS_RAW_INT
 *   30:  ATTITUDE
 *   33:  GLOBAL_POSITION_INT
 *   74:  VFR_HUD
 *   147: BATTERY_STATUS
 *
 * Each entry defines:
 *   - indices: Array positions where extracted values should be placed
 *   - parser: Function to extract values from the message payload
 */
const messageIdToIndexMap = {
  // ATTITUDE (ID 30): roll, pitch, yaw angles in radians (4 bytes each, float, little-endian)
  30: {
    indices: [0, 1, 2],
    parser: function(payload) {
      var roll  = extractFloat32LE(payload, 4);
      var pitch = extractFloat32LE(payload, 8);
      var yaw   = extractFloat32LE(payload, 12);
      return [roll, pitch, yaw];
    }
  },
  // VFR_HUD (ID 74): airspeed, groundspeed, heading, throttle
  74: {
    indices: [3, 4, 5, 6],
    parser: function(payload) {
      var airspeed    = extractFloat32LE(payload, 0);
      var groundspeed = extractFloat32LE(payload, 4);
      var heading     = extractInt16LE(payload, 8);
      var throttle    = extractUint16LE(payload, 10);
      return [airspeed, groundspeed, heading, throttle];
    }
  },
  // GLOBAL_POSITION_INT (ID 33): lat/lon in 1e-7 degrees, alt in mm
  33: {
    indices: [7, 8, 9],
    parser: function(payload) {
      var lat = extractInt32LE(payload, 4) / 1e7;
      var lon = extractInt32LE(payload, 8) / 1e7;
      var alt = extractInt32LE(payload, 12) / 1000;
      return [lat, lon, alt];
    }
  }
};

/**
 * Array to hold all parsed values.
 * Values persist between frames unless updated by new data.
 */
const parsedValues = new Array(numItems).fill(0);

//------------------------------------------------------------------------------
// Helper Functions
//------------------------------------------------------------------------------

/**
 * Extracts a 32-bit float from byte array (little-endian).
 */
function extractFloat32LE(bytes, offset) {
  var b0 = bytes[offset], b1 = bytes[offset + 1];
  var b2 = bytes[offset + 2], b3 = bytes[offset + 3];

  var sign     = (b3 & 0x80) ? -1 : 1;
  var exponent = ((b3 & 0x7F) << 1) | (b2 >> 7);
  var mantissa = ((b2 & 0x7F) << 16) | (b1 << 8) | b0;

  if (exponent === 0)
    return sign * mantissa * Math.pow(2, -149);

  if (exponent === 0xFF)
    return mantissa ? NaN : sign * Infinity;

  return sign * (mantissa | 0x800000) * Math.pow(2, exponent - 150);
}

/**
 * Extracts a 16-bit signed integer from byte array (little-endian).
 */
function extractInt16LE(bytes, offset) {
  var value = bytes[offset] | (bytes[offset + 1] << 8);
  return value > 32767 ? value - 65536 : value;
}

/**
 * Extracts a 16-bit unsigned integer from byte array (little-endian).
 */
function extractUint16LE(bytes, offset) {
  return bytes[offset] | (bytes[offset + 1] << 8);
}

/**
 * Extracts a 32-bit signed integer from byte array (little-endian).
 */
function extractInt32LE(bytes, offset) {
  var value = bytes[offset] | (bytes[offset + 1] << 8) |
              (bytes[offset + 2] << 16) | (bytes[offset + 3] << 24);
  return value > 2147483647 ? value - 4294967296 : value;
}

//------------------------------------------------------------------------------
// Frame Parser Function
//------------------------------------------------------------------------------

/**
 * Parses a MAVLink frame into the predefined array structure.
 *
 * MAVLink v1 frame structure:
 *   Byte 0:    Start marker (0xFE)
 *   Byte 1:    Payload length
 *   Byte 2:    Packet sequence
 *   Byte 3:    System ID
 *   Byte 4:    Component ID
 *   Byte 5:    Message ID
 *   Byte 6-n:  Payload
 *   Byte n+1-n+2: Checksum
 *
 * How it works:
 * 1. Validates the start marker (0xFE for v1, 0xFD for v2)
 * 2. Extracts the message ID
 * 3. Extracts the payload bytes
 * 4. Looks up the message ID in messageIdToIndexMap
 * 5. Calls the custom parser for that message type
 * 6. Stores values at their mapped indices
 *
 * @param {array} frame - The binary MAVLink frame from the data source (byte array)
 * @returns {array} Array of values mapped according to messageIdToIndexMap
 */
function parse(frame) {
  if (frame.length < 8)
    return parsedValues;

  var expectedMarker = (mavlinkVersion === 2) ? 0xFD : 0xFE;
  if (frame[0] !== expectedMarker)
    return parsedValues;

  var payloadLength = frame[1];
  var messageId = frame[5];

  var payload = [];
  for (var i = 0; i < payloadLength && (6 + i) < frame.length; i++)
    payload.push(frame[6 + i]);

  if (!messageIdToIndexMap.hasOwnProperty(messageId))
    return parsedValues;

  var config = messageIdToIndexMap[messageId];
  var values = config.parser(payload);

  for (var i = 0; i < values.length && i < config.indices.length; i++)
    parsedValues[config.indices[i]] = values[i];

  return parsedValues;
}

File: app/rcc/scripts/lua/mavlink.lua
--
-- MAVLink Messages Parser
--
-- Parses MAVLink protocol messages (drones and robotics).
--
-- INPUT FORMAT: Binary MAVLink v1 or v2 frame (byte table)
-- OUTPUT TABLE: {value1, value2, ...}  (extracted from message payloads)
--

----------------------------------------------------------------------
-- Configuration
----------------------------------------------------------------------

local numItems = 16
local mavlinkVersion = 1

local parsedValues = {}
for i = 1, numItems do parsedValues[i] = 0 end

----------------------------------------------------------------------
-- Helper Functions
----------------------------------------------------------------------

local function extractFloat32LE(bytes, offset)
  local b0 = bytes[offset]
  local b1 = bytes[offset + 1]
  local b2 = bytes[offset + 2]
  local b3 = bytes[offset + 3]

  local sign = (b3 & 0x80) ~= 0 and -1 or 1
  local exponent = ((b3 & 0x7F) << 1) | (b2 >> 7)
  local mantissa = ((b2 & 0x7F) << 16) | (b1 << 8) | b0

  if exponent == 0 then
    return sign * mantissa * 2.0 ^ (-149)
  end

  if exponent == 0xFF then
    return mantissa ~= 0 and (0/0) or sign * math.huge
  end

  return sign * (mantissa | 0x800000) * 2.0 ^ (exponent - 150)
end

local function extractInt16LE(bytes, offset)
  local value = bytes[offset] | (bytes[offset + 1] << 8)
  if value > 32767 then value = value - 65536 end
  return value
end

local function extractUint16LE(bytes, offset)
  return bytes[offset] | (bytes[offset + 1] << 8)
end

local function extractInt32LE(bytes, offset)
  local value = bytes[offset] | (bytes[offset + 1] << 8)
                | (bytes[offset + 2] << 16) | (bytes[offset + 3] << 24)
  if value > 2147483647 then value = value - 4294967296 end
  return value
end

----------------------------------------------------------------------
-- Message Parsers
----------------------------------------------------------------------

local messageParsers = {
  -- ATTITUDE (ID 30): roll, pitch, yaw in radians
  [30] = {
    indices = {1, 2, 3},
    parse = function(payload)
      return {
        extractFloat32LE(payload, 5),
        extractFloat32LE(payload, 9),
        extractFloat32LE(payload, 13),
      }
    end,
  },
  -- VFR_HUD (ID 74): airspeed, groundspeed, heading, throttle
  [74] = {
    indices = {4, 5, 6, 7},
    parse = function(payload)
      return {
        extractFloat32LE(payload, 1),
        extractFloat32LE(payload, 5),
        extractInt16LE(payload, 9),
        extractUint16LE(payload, 11),
      }
    end,
  },
  -- GLOBAL_POSITION_INT (ID 33): lat/lon in 1e-7 deg, alt in mm
  [33] = {
    indices = {8, 9, 10},
    parse = function(payload)
      return {
        extractInt32LE(payload, 5) / 1e7,
        extractInt32LE(payload, 9) / 1e7,
        extractInt32LE(payload, 13) / 1000,
      }
    end,
  },
}

----------------------------------------------------------------------
-- Frame Parser Function
----------------------------------------------------------------------

function parse(frame)
  if #frame < 8 then
    return parsedValues
  end

  local expectedMarker = mavlinkVersion == 2 and 0xFD or 0xFE
  if frame[1] ~= expectedMarker then
    return parsedValues
  end

  local payloadLength = frame[2]
  local messageId = frame[6]

  -- Extract payload bytes
  local payload = {}
  for i = 1, payloadLength do
    if 6 + i > #frame then break end
    payload[i] = frame[6 + i]
  end

  local config = messageParsers[messageId]
  if not config then
    return parsedValues
  end

  local values = config.parse(payload)
  for i = 1, #values do
    if i <= #config.indices then
      parsedValues[config.indices[i]] = values[i]
    end
  end

  return parsedValues
end
